docs/code-quality-tasks/module-graph-import-issues-report.md

```diff
@@ -0,0 +1,191 @@
+```
+
+### `packages/identity/src/ed25519/index.ts`
+- External imports (`@scure/bip39`) mixed with internal imports without proper grouping
+
+## 4. Heavy Dependencies in Utility Modules
+
+### `packages/builder/src/utils.ts`
+**Issue**: This utility module imports from `@commontools/runner`, making it a heavyweight utility rather than a lightweight one.
+
+**Impact**: 
+- Any code that needs simple utilities from `builder` must also load the entire `runner` package
+- Creates the circular dependency issue mentioned above
+- Violates the principle that utility modules should have minimal dependencies
+
+**Recommendation**: 
+- Move utilities that depend on `runner` to a separate module like `builder-runner-utils.ts`
+- Keep pure utilities without external dependencies in `utils.ts`
+
+## 5. Module-Specific Dependencies (Good Practices Found)
+
+The codebase correctly follows the guideline for module-specific dependencies:
+
+### Good Examples:
+- `packages/llm/deno.json` - Correctly declares `json5` dependency
+- `packages/ui/deno.json` - Correctly declares `@shoelace-style/shoelace` dependency
+- Most packages don't have unnecessary dependencies in their `deno.json`
+
+**This is a positive finding** - the codebase correctly avoids adding package-specific dependencies to the workspace `deno.json`.
+
+## 6. Non-standard JS Usage
+
+**Positive Finding**: The codebase is clean of non-portable JavaScript patterns:
+- No direct use of environment variables in module scope
+- No Vite-specific APIs (`import.meta.hot`, `import.meta.glob`) outside of Vite plugin code
+- Clean separation between build-time and runtime code
+
+## Recommendations
+
+### 1. Break Circular Dependencies
+Create a new package structure:
+```
+@commontools/core-types    // Shared interfaces and types
+@commontools/builder      // Depends on core-types
+@commontools/runner       // Depends on core-types and builder
+```
+
+### 2. Convert Default to Named Exports
+```typescript
+// For all identified files:
+// Change: export default Clock
+// To:     export { Clock }
+
+// Change: export default new Clock()
+// To:     export const clock = new Clock()
+//         or export function createClock(): Clock
+```
+
+### 3. Enforce Import Grouping
+Add ESLint rule or formatter configuration to enforce:
+1. Standard library imports
+2. External package imports
+3. Internal package imports
+4. Relative imports
+
+### 4. Refactor Heavy Utilities
+Split `packages/builder/src/utils.ts`:
+- `utils.ts` - Pure utilities with no external dependencies
+- `runner-utils.ts` - Utilities that need runner functionality
+
+### 5. Document Module Dependencies
+Create a module dependency graph documentation to help developers understand:
+- Which packages can be used independently
+- Which packages have peer dependencies
+- The intended hierarchy of packages
+
+## Impact Summary
+
+- **Circular dependencies**: Critical issue affecting maintainability and testability
```

Comment:
I suspect some issues we see during the vite->rollup build process (both dev and compiled) where rollup fails to find an import is due to circular dependencies, and has been popping up from seemingly benign module imports